BosunReporter/Handlers/DataDogMetricHandler.cs

```diff
@@ -0,0 +1,350 @@
+            }
+
+            _metricUri = new Uri(baseUri, "/api/v1/series");
+            _metadataUri = new Uri(baseUri, "/api/v1/metrics");
+
+            BufferWriter<byte> CreateBufferWriter() => BufferWriter<byte>.Create(MaxPayloadSize);
+
+            _metricBufferWriter = new Lazy<BufferWriter<byte>>(CreateBufferWriter);
+            _metadataBufferWriter = new Lazy<BufferWriter<byte>>(CreateBufferWriter);
+        }
+
+        /// <summary>
+        /// Constructs a new <see cref="DataDogMetricHandler" /> pointing at the specified URL
+        /// and using the given API and app keys.
+        /// </summary>
+        /// <param name="baseUrl">
+        /// URL of a DataDog endpoint.
+        /// </param>
+        /// <param name="apiKey">
+        /// An API key.
+        /// </param>
+        /// <param name="appKey">
+        /// An app key.
+        /// </param>
+        public DataDogMetricHandler(string baseUrl, string apiKey, string appKey) : this(baseUrl)
+        {
+            _metricUri = new UriBuilder(_metricUri)
+            {
+                Query = $"api_key={apiKey}"
+            }.Uri;
+
+            _metadataUri = new UriBuilder(_metadataUri)
+            {
+                Query = $"api_key={apiKey}&application_key={appKey}"
+            }.Uri;
+        }
+
+        /// <inheritdoc />
+        protected override int GetPreambleLength(PayloadType payloadType) => payloadType == PayloadType.Metadata ? 0 : s_preamble.Length;
+
+        /// <inheritdoc />
+        protected override int GetPostambleLength(PayloadType payloadType) => payloadType == PayloadType.Metadata ? 0 : s_postamble.Length;
+
+        /// <inheritdoc />
+        protected override async ValueTask SendMetadataAsync(ReadOnlyMemory<byte> buffer)
+        {
+            // metadata needs to be posted separately, one call per metric
+            // so write it into a buffer writer and continually flush it out
+            using (var bufferWriter = BufferWriter<byte>.Create())
+            {
+                foreach (var metadata in _metadata)
+                {
+                    var uri = new UriBuilder(_metadataUri)
+                    {
+                        Path = _metadataUri.AbsolutePath + "/" + metadata.Metric
+                    }.Uri;
+
+                    using (var utfWriter = new Utf8JsonWriter(bufferWriter.Writer))
+                    {
+                        JsonSerializer.Serialize(utfWriter, metadata, s_jsonOptions);
+                    }
+
+                    using (var metadataBuffer = bufferWriter.Flush())
+                    {
+                        var sequence = metadataBuffer.Value;
+                        if (sequence.IsSingleSegment)
+                        {
+                            await SendAsync(uri, HttpMethod.Put, PayloadType.Metadata, sequence.First);
+                        }
+                        else
+                        {
+                            foreach (var segment in sequence)
```

Comment:
this does *not* guarantee that each segment is valid by itself; if that is the intent, you should either use the multi-segment socket API (`IList<ArraySegment<byte>>`), or for the non-trivial case use the array pool and linearize